### PR TITLE
reactors: move into an internal package to reduce public api surface area

### PR DESCRIPTION
**Pull request by tychoish**
This patchset moves most of the core tendermint functionality into an "internal" directory to limit the exported surface area of the package, as an anticipatory move toward [stabilizing the API, as discussed in an existing ADR](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-060-go-api-stability.MD). The general belief that in order to provide a stable API, the surface of that API must be smaller than it currently is, and moving a lot of the code to the internal package is a very viable first step in this process.

This patch is aggressive! There are a couple of reasons why this is the case: first, we've come to the conclusion after talking about this internally a bunch that the reactors and the `p2p` package should be internal. Second, it's very difficult to know exactly what parts of the existing interface are actually used externally, and I think that making a bigger change will actually make it easier to use the compiler to tell us (and users!) where the missing functionality.

While this is a big change, it's also extremely not the final state of the Go API! The `node` package may need additional constructors to cover some missed use cases, and it's probably the case that some of the types in the `p2p` package will need to move into the public `types` directory. There may be additional changes that we need to make. Having said that, it's easy to add those changes in later, and managing a big fundamental change like this is very hard to if we're focused on crafting the final public API.

Additionally, given our current release schedule, if we make this change in the next week or so, then we have some time before the 0.35 release that we can use to get the API into a workable place, and then we have the entire 0.36 and 1.0 cycles that we can use to correct and polish things. The further back we push making a change like this the harder it will be. In a very exciting way, tendermint has this perfect opportunity to stabilize an API *and* be able to iterate toward stabilization in a way that allows for a good balance of "moving quickly" and also "taking enough time to potentially get it right."

So! I invite you all to review this patch set with this context, not as a final state, but as the first step in iterating toward a better API. Given the size of this PR, my intention/desire is to collect some use cases and functionality that's currently lacking and be able over the next couple of weeks.

### CODE REVIEW

**Comment by marbar3778** [comment]
Is there anyway this can be piece meal? Some of this api is currently used by users, we cant move it all.

File: internal/node/node.go [hunk 1]
```diff
@@ -0,0 +1,1354 @@
+package node
```
Comment (marbar3778) [inline internal/node/node.go:1]: parts of this needs to be in the public API. (https://github.com/cosmos/cosmos-sdk/blob/25ecec6c152aa6406da9da24151f412384a3d94b/server/start.go#L256).

Comment (tychoish) [reply internal/node/node.go:1]: The whole point of this change in general is to break the API and provide better alternatives, so this was always going to be breaking and the SDK will have to make some amount of changes at some point. 


Having said that, the code you link really looks like the sdk is duplicating the DefaultNewNode function exactly, so I actually think that the new API is the correct one. The new `node.New` just calls `DefaultNewNode` which is this: https://github.com/tychoish/tendermint/blob/node-internalize/internal/node/node.go#L93

If the SDK needs to control an aspect of the node that *isn't* controllable via the config structure, then we should change the config object to add that functionally.

Comment (marbar3778) [reply internal/node/node.go:1]: Yes it is meant to be breaking but in one fell swoop means we can be breaking things in a way that renders some apps unusable or them needing to fork the entire project. I dont think this was the goal.

**Comment by tychoish** [comment]
> Is there anyway this can be piece meal? Some of this api is currently used by users, we cant move it all.

The first couple of commits could be done separate from the last couple, but I don't think there's a functional reason to do that. Certainly, from a review perspective if you just look at the first two commits that's where anything interesting happens.

Because the interface of the legacy `node` package exports many implementation details from the reactor and p2p implementations, there's no meaningful way to preserve the legacy node package while also making any of those implementation details internal.

File: internal/node/node.go (hunk 1 above)
Comment (tychoish) [reply internal/node/node.go:1]: Is it a fair assumption that people who are using tendermint are using go mod pinned to a specific branch? This, coupled with the speed that most users tend to upgrade, would seem to give us a lot of flexibility to make changes like this without breaking anyone's apps until they intend to upgrade.  

I think the implication of the fact that the `node` package exports and explicitly depends on the interfaces of packages including all of the reactors and `p2p` means that there actually isn't a way to do this that leaves a meaningfully usable interface in more steps. We could break peoples apps in smaller ways a few times over many more releases, but I don't think that's actually a better option. 

Do you have an idea of the kind of functionality we need to provide or preserve, and examples of the projects that you think are actually depending on (internal) tendermint functionality?

Comment (cmwaters) [reply internal/node/node.go:1]: I tend to think that it's better to break in one fell swoop than over many iterations. But seeing as we kind of wanted to keep 0.35 not so breaking do you think it may be better to do these changes in 0.36. Or alternatively do a similar thing to the p2p refactor where we keep a few of the old public interfaces that people kind of need and would find annoying for us to internalise until 0.36.

Comment (tychoish) [reply internal/node/node.go:1]: > But seeing as we kind of wanted to keep 0.35 not so breaking do you think it may be better to do these changes in 0.36.

As far as I know, the go API is already expected to break somewhat in this release, and we're primarily going for an operationally simple release (e.g. no hard forks, or halts.) In a lot of ways, doing a change like this now means that we have 1-2 releases to fix anything that's really bad before 1.0, so we get a chance to stabalize.

We still may decide to push this further out, but I think it might be ok to do this now. 

> Or alternatively do a similar thing to the p2p refactor where we keep a few of the old public interfaces that people kind of need and would find annoying for us to internalise until 0.36.

I think there's less flexibility here than in the p2p because we're not expecting to keep two copies of the code active in the repo in the same way. In some ways, with API things, doing a bigger break can be easier because the compiler can enforce the rules and there isn't really the same kind of subtly. 

Let's talk about it more tomorrow/sometime.

**Comment by marbar3778** [comment]
Notes for sync later today: 

Some examples of teams using the current api: 
- https://github.com/vocdoni/vocdoni-node/search?q=tendermint%2Ftendermint
- https://github.com/vegaprotocol/wendy/blob/main/tendermint/node/node.go

Uses of types and pkgs in the sdk. 
- Reactors should be kept public to allow teams to build custom node types like the above example and https://github.com/polychainlabs/TenderSeed
- mempool interface should be kept public in order to understand how to build a custom mempool and have it work with consensus
- until we separate rpc types from internal types we have to leave some p2p and other types in the public api to allow go applications to wrap the rpc, like so https://github.com/cosmos/cosmos-sdk/blob/9fd866e3820b3510010ae172b682d71594cd8c14/client/rpc/status.go, https://github.com/cosmos/cosmos-sdk/blob/9fd866e3820b3510010ae172b682d71594cd8c14/client/broadcast_test.go
- the sdk also wraps some tendermint commands and extends them, like so https://github.com/cosmos/cosmos-sdk/blob/9fd866e3820b3510010ae172b682d71594cd8c14/x/genutil/utils.go
- Some api in tendermint is used in testing infra, like so https://github.com/cosmos/cosmos-sdk/blob/9fd866e3820b3510010ae172b682d71594cd8c14/testutil/network/util.go

File: internal/node/node.go (hunk 1 above)
Comment (cmwaters) [reply internal/node/node.go:1]: > I think there's less flexibility here than in the p2p because we're not expecting to keep two copies of the code active in the repo in the same way

I haven't looked into your implementation in any depth (going to do that now) but I would imagine the public API's that you might want to keep would just point directly to the internal ones so you're not really keeping two copies of everything.

> we're primarily going for an operationally simple release

yeah good point

File: node/node.go [hunk 2]
```diff
@@ -1,1356 +1,15 @@
...
-func getChannelsFromShim(reactorShim *p2p.ReactorShim) map[p2p.ChannelID]*p2p.Channel {
-	channels := map[p2p.ChannelID]*p2p.Channel{}
-	for chID := range reactorShim.Channels {
-		channels[chID] = reactorShim.GetChannel(chID)
-	}
-
-	return channels
+// New constructs a tendermint node service for use in go process that
+// host their own process-local tendermint node.
+func New(conf *config.Config, logger log.Logger) (service.Service, error) {
```
Comment (cmwaters) [inline node/node.go:13]: Having these arguments to set up a node means that users that run Tendermint programatically can't use custom-reactors (or a different state-provider but that's probably less of a concern). I'm not saying that we shouldn't want this, I just think we need to be deliberate in the decision when it comes to allowing / disallowing custom reactors

Comment (cmwaters) [reply node/node.go:13]: Perhaps more important than this is that it looks like with this configuration that nodes won't be able to run Tendermint in-process. We should have something like  a `proxy.ClientCreator` or a way of optionally passing the `abci.Application` interface when applications aren't using a socket.

File: internal/node/node.go (hunk 1 above)
Comment (tychoish) [reply internal/node/node.go:1]: > I would imagine the public API's that you might want to keep would just point directly to the internal ones so you're not really keeping two copies of everything.

Totally, I think between this commit (potentially) landing and 0.35 we'll definitely move some things out of `internal` and into the new `node` package (and `types` etc,) to facilitate some this. 

--- 

On a related note, and this might be subtle, so I'll underscore it: the `Node` constructors now return a `service.Service` interface, which means users (even internal) won't be able to (easily) reach into the node instances itself which will make simplification and will ease using these types externally.

**Review by cmwaters — COMMENTED** [review]
I haven't gone through everything yet but I have added a few comments.

One other thing is that with this setup we treat node as a `service.Service` which I think is not bad but I would think that users running Tendermint programatically might want to not just start and stop the service but stop it, change the config and start it again without having to reinitiate everything.

**Comment by cmwaters** [comment]
I think we should also discuss internalising parts (or most) of the `proxy` and `state` directories

**Comment by tychoish** [comment]
I've updated the description of this with a summary of the changes, which I think may be able to more effectively guide any kind of review.

File: proto/tendermint/consensus/types.pb.go [hunk 3]
```diff
@@ -1,17 +1,18 @@
 // Code generated by protoc-gen-gogo. DO NOT EDIT.
-// source: tendermint/consensus/types.proto
```
Comment (marbar3778) [inline proto/tendermint/consensus/types.pb.go:2]: these are auto-generated files, we should not edit them. gofmt is ignored in these files as well. Can we revert these changes.

Comment (tychoish) [reply proto/tendermint/consensus/types.pb.go]: done!

**Comment by ValarDragon** [comment]
I get the p2p work being moved to internal, but I don't think the `netaddress`, `addressbook`, or `mempool` code should be moved to internal. These are things that it is quite natural for tooling outside of Tendermint to have to interface with. E.g. other tools that want to automate or format peer addresses.

**Comment by tychoish** [comment]
Thanks @ValarDragon for your feedback/thoughts! 

> I get the p2p work being moved to internal, but I don't think the netaddress, addressbook, or mempool code should be moved to internal.

The `p2p.AddrBook` is part of the legacy `p2p` stack, and is going to be removed immediately following the 0.35 release, so I'm not sure that it makes sense to pull it out of the internal package only to have it be deleted.

I think the process of meaningfully including the mempool as part of the public interface is a thing that we can definitely talk about, but one that's much larger than this PR, and involves some amount of though around what the use cases are for user-supplied mempools, how the rest of the system needs to tolerate (or not) variances in mempool behavior and semantics, and some design for what a reasonable interface for a mempool is. These are great questions and discussions to have, but this PR doesn't preclude this, and indeed makes it easier to have these conversations in the future because we won't need to ratify or work around any of the existing interfaces. 

I believe that the netaddress makes sense to move into `types` in the short term (e.g. before 0.35), but I think it would be better to do that in it's own PR separate from this, because I think it's better to keep this PR simple (even if it's large.)

**Comment by tychoish** [comment]
Ok, I've rebased this on top master, which now include #6540 and #6533, which makes this change *only* moving files and fixing import statements.

**Comment by tychoish** [comment]
I'm going to close this and attempt to accomplish something similar via other means
